src/platform/backends/lxd/lxd_vm_image_vault.cpp

```diff
@@ -0,0 +1,216 @@
+
+        auto json_reply =
+            lxd_request(manager.get(), "POST", QUrl(QString("%1/images").arg(base_url.toString())), image_object);
+
+        if (json_reply["metadata"].toObject()["class"] == QStringLiteral("task") &&
+            json_reply["status_code"].toInt(-1) == 100)
+        {
+            QUrl task_url(QString("%1/operations/%2")
+                              .arg(base_url.toString())
+                              .arg(json_reply["metadata"].toObject()["id"].toString()));
+
+            // Instead of polling, need to use websockets to get events
+            while (true)
+            {
+                try
+                {
+                    auto task_reply = lxd_request(manager.get(), "GET", task_url);
+
+                    if (task_reply["error_code"].toInt(-1) != 0)
+                    {
+                        break;
+                    }
+
+                    auto status_code = task_reply["metadata"].toObject()["status_code"].toInt(-1);
+                    if (status_code == 200)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        auto download_progress = parse_percent_as_int(
+                            task_reply["metadata"].toObject()["metadata"].toObject()["download_progress"].toString());
+
+                        if (!monitor(LaunchProgress::IMAGE, download_progress))
+                        {
+                            lxd_request(manager.get(), "DELETE", task_url);
+                            throw mp::AbortedDownloadException{"Download aborted"};
+                        }
+
+                        std::this_thread::sleep_for(5s);
+                    }
+                }
+                // Implies the task is finished
+                catch (const LXDNotFoundException&)
+                {
+                    break;
+                }
+            }
+        }
+    }
+
+    return source_image;
+}
+
+void mp::LXDVMImageVault::remove(const std::string& name)
+{
+    try
+    {
+        lxd_request(manager.get(), "DELETE",
+                    QUrl(QString("%1/virtual-machines/%2").arg(base_url.toString()).arg(name.c_str())));
+    }
+    catch (const LXDNotFoundException&)
+    {
+        mpl::log(mpl::Level::warning, category, fmt::format("Instance \'{}\' does not exist: not removing", name));
+    }
+}
+
+bool mp::LXDVMImageVault::has_record_for(const std::string& name)
+{
+    try
+    {
+        lxd_request(manager.get(), "GET",
+                    QUrl(QString("%1/virtual-machines/%2").arg(base_url.toString()).arg(name.c_str())));
+
+        return true;
+    }
+    catch (const LXDNotFoundException&)
+    {
+        return false;
+    }
+}
+
+void mp::LXDVMImageVault::prune_expired_images()
+{
+}
+
+void mp::LXDVMImageVault::update_images(const FetchType& fetch_type, const PrepareAction& prepare,
+                                        const ProgressMonitor& monitor)
+{
+}
```

Comment:
These are TODO, right? We should definitely prune old images, and {c,sh}ould we drive the update some way?

Reply:
They _might_ be TODO.  LXD can also handle this in and of itself, so should we config LXD to do this for us or we drive it?

Reply:
If we can make it DTRT without our involvement, probably best.

Reply:
Ok, then I will look into that and failing being able to have LXD control all of that, I'll implement it here.